### PR TITLE
Withdraw tasker workflow and interaction with smart contract

### PR DESCRIPTION
**Pull request by jennsun**
with @christine-sun 

Implemented the tasker's workflow to withdraw funds from a task. 
- Created withdraw pages for each specific task, displaying task information and payment amount
- Updated statuses to reflect new status map as follows:
0: in progress
1: pending approval
2: ready to withdraw
3: abandoned
4: paid
5: rejected
- Rendered withdraw/abandon buttons according to proper status
- Withdraw only after tasker's task has been approved by reviewer after @bzzbbz changes
- Display loading status as withdraw transaction is being mined
- Withdraw the proper price per task amount from the smart contract for the task in WMATIC to the user
- Update status of task after withdrawing to indicate the task has been "Paid"

https://user-images.githubusercontent.com/26675662/165890736-6bfb4fb2-3b51-4ad7-85a0-c5f55287afe1.mp4

### CODE REVIEW

File: pages/requester/create-task.tsx [hunk 1]
```diff
@@ -134,52 +138,147 @@ export default function Form() {
...
           signer
         );
+        console.log("2");
         // Get reference to MATIC Token contract
         const maticContract = new ethers.Contract(
           maticTokenAddress,
           ERC20ABI,
           signer
         );
+        console.log("3");
```
Comment (bzzbbz) [inline pages/requester/create-task.tsx:260]: remove log

File: pages/requester/create-task.tsx [hunk 2]
```diff
@@ -134,52 +138,147 @@ export default function Form() {
...
         var bigNumCryptoAllocated = BigNumber.from(
           parseFloat(cryptoAllocated) * 10 ** 8
         ).mul(BigNumber.from(10).pow(10));
 
         // Approve accessed to user's WMATIC
+        console.log("4");
```
Comment (bzzbbz) [inline pages/requester/create-task.tsx:266]: remove log

File: pages/requester/create-task.tsx [hunk 3]
```diff
@@ -134,52 +138,147 @@ export default function Form() {
...
 
         // Fund the contract
+        console.log("5");
         const escrowFundTxn = await escrow.fund(
           bigNumCryptoAllocated.toString()
         );
 
         return { contractAddress: newContractAddress, error: null };
+        await escrowFundTxn.wait();
+        console.log("6");
```
Comment (bzzbbz) [inline pages/requester/create-task.tsx:281]: remove log

File: components/task/TaskWithdraw.tsx [hunk 4]
```diff
@@ -0,0 +1,148 @@
+import { Typography, Stack, Grid, Box, CircularProgress } from "@mui/material";
+import PrimaryButtonCTA from "../../components/buttons/PrimaryButtonCTA";
+import { TaskOverviewData } from "../../src/Types";
+import Escrow from "../../src/utils/abi/Escrow.json";
+import { ethers } from "ethers";
+import { CreatedTaskStatus, TaskData } from "../../src/Types";
```
Comment (bzzbbz) [inline components/task/TaskWithdraw.tsx:6]: unused import

File: pages/requester/create-task.tsx [hunk 5]
```diff
@@ -20,12 +20,15 @@ import { ethers, BigNumber } from "ethers";
...
+import MulticallABI from "../../src/utils/abi/Multicall.json";
 
 export default function Form() {
-  const escrowFactoryAddress = process.env
-    .NEXT_PUBLIC_ESCROW_FACTORY_ADDRESS as string;
-  const maticTokenAddress = process.env
-    .NEXT_PUBLIC_MATIC_TOKEN_ADDRESS as string;
+  // const escrowFactoryAddress = process.env
+  //   .NEXT_PUBLIC_ESCROW_FACTORY_ADDRESS as string;
+  // const maticTokenAddress = process.env
```
[the remark this answers is not in the document]
Comment (nicholaspad) [reply pages/requester/create-task.tsx:28]: Yea, we should prefer using env vars here

File: pages/requester/create-task.tsx [hunk 6]
```diff
@@ -295,7 +394,7 @@ export default function Form() {
               <Grid item xs={6}>
                 <Box display="flex" alignItems="center" justifyContent="left">
                   <Typography color="primary" sx={{ textAlign: "right" }}>
-                    Total ETH allocated:
+                    Total MATIC allocated:
```
Comment (nicholaspad) [inline pages/requester/create-task.tsx:397]: I would do a global search for "ETH" and replace as many places as possible with "MATIC"

File: cloud/cloud.js [hunk 7]
```diff
@@ -856,17 +858,34 @@ Moralis.Cloud.define(
...
+    if (!res) return;
+
+    res.set("status", 4);
+    await res.save();
   }
-);
+
+  const taskId = request.params.taskId;
+  const ethAddress = request.user.get("ethAddress");
+
```
Comment (nicholaspad) [inline cloud/cloud.js:881]: Before calling `updateTaskStatus`, there should be checks:

1. Check that the tasker has actually claimed the task

```javascript
if (!(await checkTaskerClaimedTask(ethAddress, taskId)))
      return {
        success: false,
        message: `Address ${ethAddress} has not claimed task ${taskId}.`,
      };
```

2. Check that the current TaskUsers `status` value is `2` (ready to withdraw), and if not, do not proceed

File: cloud/cloud.js [hunk 8]
```diff
@@ -856,17 +858,34 @@ Moralis.Cloud.define(
...
+    const TaskUsers = Moralis.Object.extend(tableName);
+    const query = new Moralis.Query(TaskUsers);
+
+    const res = await query
+      .equalTo("taskerId", taskerId)
+      .equalTo("taskId", taskId)
+      .first();
+
+    if (!res) return;
+
```
Comment (nicholaspad) [inline cloud/cloud.js:874]: Check that current status is `2`:

```javascript
if (res.get("status") !== 2) return
```

File: pages/requester/create-task.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (jennsun) [reply pages/requester/create-task.tsx:28]: going to keep for now -- was during the multicall process

**Review by nicholaspad — CHANGES_REQUESTED** [review]

**Review by nicholaspad — APPROVED** [review]

**Review by bzzbbz — APPROVED** [review]

**Pull request merged**